Typescript development in Cirq and prototype bloch sphere implementation

seunomonije opened the pull request:
### Typescript Development in Cirq
This PR implements the features discussed in the [Typescript Development in Cirq RFC](https://docs.google.com/document/d/16LybO9bNqL3LaIy1pZCJQXSPvKs5g3IqKkkE3efqM8U/edit). 

#### Listed below are the features included in the PR:
 - Infrastructure and directory structure to allow Typescript development in Cirq, including:
   - Bundling Typescript into Javascript, and examples of how to use these bundles
   - Hot reloading development environment for both browser and notebook settings
   - Typescript testing framework using Mocha and Chai, including a test for the bloch sphere.
   - Coverage checks for Typescript unit tests. (100% coverage needed in CI)
   - End-to-end integration between the Python and Javascript (the bloch sphere)
   - Example of full coverage testing of the Python side
   - CI for the Typescript environment for testing and linting. 
   - Check scripts for testing and linting
 - An example 3d bloch sphere implementation. This bloch sphere can be viewed:
   - In the browser (development)
   - In a Jupyter notebook (development and production)
   - Via a generated HTML file. This file can be sent anywhere and be viewed in the browser, regardless of if the recipient has Cirq installed or not. 
 - A [up-to-date package on Test PyPI](https://test.pypi.org/project/cirq-web/0.12.0.dev20210616161134/) representing the state of the cirq-web package as of 6/16/21.
 - Documentation for using cirq-web in the README files.

![colab_web_pr_trial](https://user-images.githubusercontent.com/17647506/122309538-1673ab80-ced4-11eb-96dd-112ea46b6df9.gif)

balopat reviewed (changes requested):
Good stuff, I'm very excited about this :) 

Here's my first round of review! Added a bunch of comments and questions.

d.ts files - how are we planning to keep them in sync? Do we really need them?

I will add more comments soon.

seunomonije commented on cirq-web/setup.py:
```diff
@@ -0,0 +1,69 @@
+# This is a pure metapackage that installs all our packages
+requirements += [f'cirq-core=={__version__}']
+
+packs = ['cirq_web']
```
Sounds good. I think packaging is something I'll handle right before we merge in the PR.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/CirqBlochSphere.class.ts:
```diff
@@ -0,0 +1,63 @@
+import {Group} from 'three';
+
+export class CirqBlochSphere {
+  RADIUS: number;
```
Good catch

seunomonije commented on cirq-web/.gitignore:
```diff
@@ -0,0 +1,9 @@
+.pytest_cache/
+
+# Extras
+utils/
```
At first it was for my helper scripts and things like that, but after I saw the dev scripts you implemented in Cirq (like `check/`), I thought it may be helpful to have a similar folder that helps with things like updating the `webpack.config.js` file automatically, stuff like that. Thinking more about it though it may be better just to have users enter the data into the config.js file themselves (I gave instructions in the README already)...there's nothing complex enough that warrants script help yet. However, its something that's pretty cool and if we get to the point where we do find something complex enough I'd want to try and implement it. Knowing what I know now, it probably will need to go in the root directory of Cirq anyway and not in `cirq-web`.

TLDR: Moving it out and deleting from gitignore.

seunomonije commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,77 @@
+developed in general. In the general sense, visualizations should have:
+ - A "root" folder within the `src/` directory labeled according to the title of the visualization. All files and directories for a particular visualization will live here. In the case of the Bloch Sphere, this is `bloch_sphere/`.
+ - A `components/` directory which contains classes representing different components of the larger visualization, following typical object oriented programming techniques. In the case of the Bloch sphere, you can see that we have different classes for `Axes`, `Meridians`, `Text` etc.
+ - Any additional directories with information necessary for the visualization (fonts, images, etc.). In the case of the Bloch Sphere, we can see a `fonts/` directory which holds necessary font data. 
```
I like the an `assets/` directory idea a lot..allows for the case you said and for people to have even deeper nested directories for things like fonts. For the example I'd want to keep the structure neat and clean at least, but I'll also specify in the README that its loose.

seunomonije commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,77 @@
+ - A "root" folder within the `src/` directory labeled according to the title of the visualization. All files and directories for a particular visualization will live here. In the case of the Bloch Sphere, this is `bloch_sphere/`.
+ - A `components/` directory which contains classes representing different components of the larger visualization, following typical object oriented programming techniques. In the case of the Bloch sphere, you can see that we have different classes for `Axes`, `Meridians`, `Text` etc.
+ - Any additional directories with information necessary for the visualization (fonts, images, etc.). In the case of the Bloch Sphere, we can see a `fonts/` directory which holds necessary font data. 
+ - A larger class within the visualization's "root" folder which brings the individual components of the visualization together. In the case of the Bloch Sphere, this is `CirqBlochSphere.class.ts`.
```
I don't think its important actually, when initially toying around with everything we started it that way and it just stuck. I'm going to go through and rename things anyways, delete the `*.d.ts` files, make filenames lowercase, etc, so this will definitely change.

seunomonije commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,77 @@
+
+### TODO: Testing Visualizations
+
+- **Testing, which framework?-- important**
```
Just set up a framework with Mocha and Chai lol. Three.js uses Mocha in their documentation so decided to go that route, and everything is working straightforward so I think we might just stay that way. I was looking at some other open source Google projects and they use Mocha as well so I think we should be good...but always open to change if there are features that make Jasmine better

seunomonije commented on cirq-web/cirq_ts/dist/index.html:
```diff
@@ -0,0 +1,13 @@
+  <div id="container"></div>
+  <script src="bloch_sphere.bundle.js"></script>
+  <script>
+    createSphere.showSphere('{"radius": 5}');
```
Ah forgot about this..in the `webpack-config-js` file we have the output library name set to "createSphere", but we probably should set it to something more general like even "cirq_web" or something like that. I'll also need to change the showSphere to showBlochSphere or something similar. 

I'll do this after I finish all the other comments since it requires rebundling and changing HTML both on the Python and TS side.

seunomonije commented on cirq-web/cirq_ts/package.json:
```diff
@@ -0,0 +1,42 @@
+    "start": "webpack serve --mode development",
+    "build": "npm test && webpack --mode production",
+    "test": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' mocha -r ts-node/register 'test/**/*.ts'",
+    "lint": "gts lint",
```
It may be the lint configuration or maybe some manual error. I ran it a few times and it auto fixed files `npm run fix`. I had it down to where there were no errors except one complaining that three.js was an unpublished package (I think the rule was `no-unpublished-package`?). Going to look into that later though and also maybe check the config to make sure everything is in line.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/CirqBlochSphere.class.d.ts:
```diff
@@ -0,0 +1,13 @@
+import { Group } from 'three';
+export declare class CirqBlochSphere {
```
Don't think they're needed actually.. tsc automatically generated them so I thought they would be good to have, but also after looking into it there's no real use to them for our purposes.

Going to delete them.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/CirqBlochSphere.class.ts:
```diff
@@ -0,0 +1,63 @@
+    this._init();
+  }
+
+  public createSphere() {
```
Renamed to `getBlochSphere()`. Thoughts?

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/CirqBlochSphere.class.ts:
```diff
@@ -0,0 +1,63 @@
+    this._addHorizontalMeridians();
+    this._addVerticalMeridians();
+    this._addAxes();
+    this._addAxes();
```
It is, good catch.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Axes.class.ts:
```diff
@@ -0,0 +1,46 @@
+    let geometry: BufferGeometry;
+    geometry = new BufferGeometry().setFromPoints(xAxis);
+    const blueLine = new Line(
+      geometry,
```
Taken care of in your next comment

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Axes.class.ts:
```diff
@@ -0,0 +1,46 @@
+    const xAxis = [new Vector3(-radius, 0, 0), new Vector3(radius, 0, 0)];
+    const yAxis = [new Vector3(0, 0, -radius), new Vector3(0, 0, radius)];
+    const zAxis = [new Vector3(0, -radius, 0), new Vector3(0, radius, 0)];
+    const colors = {
```
No access, I requested it so hopefully I get some soon

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Axes.class.ts:
```diff
@@ -0,0 +1,46 @@
+      yAxis: '#ff3131', // neon red
+      zAxis: '#39ff14', // neon green
+    };
+    const lineWidth = 1.5;
```
I think so. IMO, the only thing that should be configurable are the colors of the axes, but that's also something that's up for discussion.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Meridians.class.ts:
```diff
@@ -0,0 +1,146 @@
+      const distance2 = Math.pow(position, 2);
+      const newRadius = Math.sqrt(hyp2 - distance2); //radius^2 - b^2 = a^2
+
+      Meridians._curveData.radius = newRadius;
```
Will do. Like this solution

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Meridians.class.ts:
```diff
@@ -0,0 +1,146 @@
+  public static createHorizontalCircleMeridians(radius: number): Line[] {
+    Meridians._curveData.radius = radius;
+    const meridians = [];
+    for (let i = 0; i < Math.PI; i += Math.PI / 4) {
```
Will come back to this because I think I can use it elsewhere too..can also chat about it later.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Meridians.class.ts:
```diff
@@ -0,0 +1,146 @@
+   * rendered by the three.js scene.
+   * @param curve An EllipseCurve object that provides location/size info
+   * @param rotationFactor The desired angle of rotation in radians
+   * @param vertical (Optional) boolean that tells whether or not we're generating a horizontal
```
Done

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Meridians.class.ts:
```diff
@@ -0,0 +1,146 @@
+import {EllipseCurve, BufferGeometry, LineBasicMaterial, Line} from 'three';
+
+export class Meridians {
+  private static _curveData = {
```
The purpose of the fns here is to output an array of `Line` objects that can easily be added to the three.js bloch sphere `Group`..are you suggesting that I wrap the output in a `Meridian` type with the `Line` info inside? IMO it's clear just to output the line objects and add..on the other end we'd have to unpack them. It's not that big of a deal but just seems like an extra step. I can see tho how it would be clearer for organization.

What do you think?

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Scene.class.ts:
```diff
@@ -0,0 +1,106 @@
+   */
+  private addSceneToHTML() {
+    // We must assume that there is already a div element named 'container'
+    const container = document.getElementById('container')!;
```
Will take care of this after other changes.

seunomonije replied:
Commit 2527c57 allows users to pick the div id where the viz output goes, but thinking more about how people will be using the bloch sphere, will there ever be a need for multiple bloch spheres? I can see maybe when wanting to compare bloch spheres in development, but in a notebook setting things might get crowded, unless I develop some HTML file(s) that organize bloch spheres based on the numbers they output, scale the scene sizes and everything else accordingly, etc. And all that will need to be specified from Python, which could add some more work on the users part. Sometime I think to look at for the future possibly. 

Thoughts?

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Scene.class.ts:
```diff
@@ -0,0 +1,106 @@
+   * the 'container' div.
+   */
+  private addSceneToHTML() {
+    // We must assume that there is already a div element named 'container'
```
Will be changed when scaling for multiple bloch spheres

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Vector.class.ts:
```diff
@@ -0,0 +1,56 @@
+   * @param vectorData information representing the location of the vector
+   * @returns an ArrowHelper object to be rendered by the scene.
+   */
+  public static createVector(vectorData?: string): ArrowHelper {
```
I think this goes back to the recurring issue where I call static methods from a class..what I can do is create a Vector type and pass it in there. The reason why I have this as JSON is organizationally the two elements that require input from the user both can take JSON in the main method so its straightforward where everything is going.

What do you think about making a Vector class that gathers/parses json data in the constructor and then we can call a method createVector with a new Vector type, like I do for the `BlochSphere` class?

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Vector.class.ts:
```diff
@@ -0,0 +1,56 @@
+   * @returns an ArrowHelper object to be rendered by the scene.
+   */
+  public static createVector(vectorData?: string): ArrowHelper {
+    let inputData;
```
See above, can be vector type

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/main.ts:
```diff
@@ -0,0 +1,15 @@
+
+  const scene = new BlochSphereScene();
+
+  const object = new CirqBlochSphere(inputData.radius).createSphere();
```
I added a default of radius 5 to the constructor

seunomonije commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,77 @@
+### Developing in Jupyter Notebook
+An alternative to developing using `webpack-dev-server` is to bundle the Typescript and reference the Javascript output. You can spin up a notebook server with `jupyter notebook`, and bundle (while watching for live changes and updating accordingly) with `npx webpack --mode production --watch`. These processes must run simultaneously. This is especially useful for if you want to work with integra ...
+
+**NOTE:** In order to access the bundled javascript, you need to include the full path to it: `cirq_ts/dist/YOUR_VIZ_NAME.bundle.js`. If you make any changes to the directory structure, take into account that the path may change as well.
```
I see this name as a placeholder for now...when the actual PR goes through the big review we will probably be renaming the whole thing. Something to bring up again in Cirq Sync.

balopat replied:
We need that prefix, because that will be the actual root level package name in the installation folder, e.g. `<python-env>/lib/site-packages/cirq_ts` and `<python-env>/lib/site-packages/cirq_web` - otherwise it might collide with something called `ts` or `web`, and also then the user will have to say `import web` or `import ts` (although the latter won't be necessary).

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Text.class.ts:
```diff
@@ -0,0 +1,56 @@
+      // explicitly typing so we can access later
+      '|+>': new Vector3(5, 0, -0.1), // z proportional to the height
+      '|->': new Vector3(-5 - labelSize, 0, -0.1),
+      'i|->': new Vector3(0, 0, 5),
```
The ⟩ character doesn't render in three.js. Looking into alternatives.

seunomonije replied:
Issue solved yesterday.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/Axes.class.ts:
```diff
@@ -0,0 +1,46 @@
+    const xAxis = [new Vector3(-radius, 0, 0), new Vector3(radius, 0, 0)];
+    const yAxis = [new Vector3(0, 0, -radius), new Vector3(0, 0, radius)];
+    const zAxis = [new Vector3(0, -radius, 0), new Vector3(0, radius, 0)];
+    const colors = {
```
My understanding is that these axes should be a standard color across all bloch spheres. (Eg. x is blue, y is red, z is green) so that the user has an additional indicator when comparing states. In what cases do you think they should be configurable? Or by configurable do you mean able to be changed within the code easier, but to whatever color we determine?

balopat commented on .github/workflows/ci.yml:
```diff
@@ -266,3 +266,26 @@ jobs:
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          # Do we want check multiple versions of Node.js?
```
no, we don't, this is development only, we control the dev env. If the nodejs side would be something that we think is going to be used as a library, then we would - but that is not the case!

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/scene.ts:
```diff
@@ -0,0 +1,122 @@
+   * renderer
+   * controls - the Three.js OrbitControls object reponsible for managing
+   *  mouse input and moving the camera, zooming in and out, etc.
+   * divId - the id of the HTML div that will contain the scene output
```
This is not necessarily a div, it can be any container - `<div>` or `<span>` typically but maybe even `<body>`...I'd suggest renaming to `containerId`

balopat reviewed (commented):
Starting to shape up quite nicely. I'm going to continue tomorrow the review.

seunomonije commented on cirq-web/requirements.txt:
```diff
@@ -0,0 +1 @@
+IPython
```
The only rationale I have for including it is that in a notebook users can use `display()` without actually needing to add IPython themselves. Looking back at it though I can see how it would be almost equally as easy for people to just import IPython and do it themselves. Changing now.

balopat replied:
Yeah, for notebook users, the notebook environment already typically includes IPython, as notebook is just a web version of IPython.

balopat commented:
@googlebot please rescan

seunomonije commented:
This most recent commit should fail TS-coverage CI, there's something with the tool that's reporting missed branch coverage on the constructor in `bloch_sphere.ts` when I initialize it with a specified radius value. Looking into why this issue is happening.

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/test_bloch_sphere.ts:
```diff
@@ -0,0 +1,61 @@
+
```
Add standard copyright notice.

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/test_bloch_sphere.ts:
```diff
@@ -0,0 +1,61 @@
+  
+  it('Bloch sphere with |0⟩ statevector is correct', () => {
+    const expected = PNG.PNG.sync.read(readFileSync('e2e/bloch_sphere/bloch_sphere_expected.png'))
+    const actual = PNG.PNG.sync.read(readFileSync('e2e/bloch_sphere/bloch_sphere.png'))
```
We should write this conditioned on an environment variable to either a temporary directory, or to the `bloch_sphere_expected.png`

seunomonije replied:
I have it in a temp directory, but wdym by conditioned on an environment variable ? Ie, user sets whether or not they want to have a file generated, and based on that setting, it will build it in a temp or not? Included in the commit that fixes this, I have a `utils/` folder that allows the user to generate a current screenshot of the bloch sphere via the current bundle.

seunomonije commented:
TS unit and TS covg are failing due to an issue with the image diffs. Pass on my machine, but fail online. I wonder if unicode characters have anything to do with it..this next commit will test it out.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/main.ts:
```diff
@@ -0,0 +1,43 @@
+ * @param divId A string containing the div id that will contain the visualization
+ * output.
+ */
+export function showSphere(
```
```suggestion
export function blochSphere(
```

wcourtney replied:
Nit: functions (especially those with no return value) should describe the side effect that they make, which should be something like an imperative clause.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/main.ts:
```diff
@@ -0,0 +1,43 @@
+ * bloch sphere.
+ * @param vectorData A JSON string containing information that configures the
+ * state vector.
+ * @param divId A string containing the div id that will contain the visualization
```
```suggestion
 * @param containerId A string containing the container (div, span, etc.) id that will contain the visualization
```

balopat commented on cirq-web/cirq_ts/tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    "downlevelIteration": true,
+    "declaration": false,
+    "sourceMap": false,
+    //"rootDir": ".",
```
do we need this?

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,140 @@
+
+        return path_of_html_file
+
+    def _convertSphereInput(self, radius):
```
use snake_case instead of CamelCase in function names

balopat replied:
type hints for args and return type are missing

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,140 @@
+        obj = {'radius': radius}
+        return json.dumps(obj)
+
+    def _createVector(self, state_vector):
```
use snake_case instead of CamelCase in function names

balopat replied:
type hints for args and return type are missing

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,140 @@
+
+        return bloch_vector
+
+    def _createRandomVector(self):
```
use snake_case instead of CamelCase in function names

balopat replied:
type hints for args and return type are missing

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,140 @@
+        bloch_vector = bloch_vector_from_state_vector(state_vector, 0)
+        return bloch_vector
+
+    def _serializeVector(self, x, y, z, length=5):
```
use snake_case instead of CamelCase in function names

balopat replied:
type hints for args and return type are missing

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,140 @@
+        return bloch_vector
+
+    def _serializeVector(self, x, y, z, length=5):
+        # .item() bc input is of type float32, need to convert to seralizable type
```
```suggestion
        # .item() because input is of type float32, need to convert to serializable type
```

Also, can we try whether `cirq.to_json()` would work on the vector?

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,140 @@
+        """
+
+    def generate_HTML_file(
+        self, output_directory='./', file_name='bloch_sphere.html', open_in_browser=False
```
type hints for args and return type are missing

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere_test.py:
```diff
@@ -0,0 +1,117 @@
+from cirq.qis.states import bloch_vector_from_state_vector
+
+
+def get_bundle_file_path():
```
We typically (although I'm not sure it's consistent across the codebase) make methods private (`_get_bundle_file_path`) in tests unless we want them to be reused across different test files - but then we typically add them to `cirq.testing` as a test utility and write tests for them too.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere_test.py:
```diff
@@ -0,0 +1,117 @@
+
+@pytest.mark.parametrize('sphere_radius', [5, 0.2, 100])
+def test_valid_bloch_sphere_radius_json_info(sphere_radius):
+    bloch_sphere = cirq_web.BlochSphere(sphere_radius=sphere_radius)
```
we should hook up cirq json serialization instead of rolling our own, so that cirq.to_json() and cirq.from_json() works well with these items as well.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere_test.py:
```diff
@@ -0,0 +1,117 @@
+    bloch_sphere = cirq_web.BlochSphere()
+    bundle_script = get_bundle_file_path()
+    expected = f"""
+        <meta charset="UTF-8">
```
I'm not sure this is needed in every cell in a notebook, or is it?

seunomonije replied:
My stance on this is i'd rather have it there just in case, than someone come across an error and spend a lot of time going through everything because they don't have it.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere_test.py:
```diff
@@ -0,0 +1,117 @@
+    assert expected == bloch_sphere._repr_html_()
+
+
+def test_generate_HTML_file_with_browser(tmpdir):
```
```suggestion
def test_generate_html_file_with_browser(tmpdir):
```

nit: let's use consistent casing for html - I think `html` is fine

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,103 @@
+
+
+def determine_env():
+    """Determines if a Widget is being run in a Jupyter notebook
```
```suggestion
    """Determines if a Widget is being run in a Jupyter notebook.
```

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,103 @@
+    elif env == 'google.colab_shell':
+        return Env.COLAB
+    else:
+        return Env.OTHER
```
we can even have this as a defaultdict

```python
_ENV_MAP: DefaultDict[str, Env] =defaultdict(lambda: Env.OTHER, {
    'ZMQInteractiveShell': Env.JUPYTER,
    'google.colab_shell': Env.COLAB,
})

def determine_env():
     return _ENV_MAP[IPython.get_ipython().__class__.__name__]
```

seunomonije replied:
Like this. Implementing now.

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,103 @@
+
+        file_name: the name of the output file. Default is 'bloch_sphere'
+
+        contents: the contents of the file
```
missing `Returns` 
+ type hints

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,103 @@
+    """Parent class for all widgets."""
+
+    def __init__(self, bundle_file_path):
+        """Initializes a Widget, gathering it's respective bundle file path and
```
```suggestion
        """Initializes a Widget.
        
         During initialization it gathers its bundle file path and
```

balopat replied:
+ type hints

balopat commented on cirq-web/cirq_web/README.md:
```diff
@@ -0,0 +1,80 @@
+
+### Handling HTML output from Python
+#### Viewing a visualization in a notebook setting
+We capitalize on IPython's `_repr_html_` magic method to help display visualiztions in the notebook. In the main class of your visualization, include a method like so:
```
```suggestion
We capitalize on IPython's `_repr_html_` magic method to help display visualizations in the notebook. In the main class of your visualization, include a method like so:
```

balopat reviewed (changes requested):
This is becoming very nice :) I only had nits for now and maybe trying to leverage cirq.to_json / from_json to work. Please also do merge master, there will be some simplifications to the top level ./setup.py too!

wcourtney commented on cirq-web/cirq_ts/src/bloch_sphere/components/scene.ts:
```diff
@@ -0,0 +1,137 @@
+   * @param height The desired height of the rendering canvas.
+   */
+  public setRenderSize(
+    width: number = BlochSphereScene.VIZ_WIDTH,
```
It's a little odd to call `setRenderSize()` with no arguments since it's not clear in that context what the size will be. Consider ditching the default values unless I'm missing something?

wcourtney commented on cirq-web/cirq_ts/src/bloch_sphere/components/vector.ts:
```diff
@@ -0,0 +1,71 @@
+
+/**
+ * Adds a state vector to the bloch sphere.
+ * @param vectorData information representing the location of the vector.
```
@param and @returns is for a function; no need for those annotations here.

seunomonije replied:
Sloppy on my part..forgot to move

wcourtney commented on cirq-web/cirq_ts/src/bloch_sphere/components/vector.ts:
```diff
@@ -0,0 +1,71 @@
+  x: number;
+  y: number;
+  z: number;
+  v_length: number;
```
If this is the vector length, then recommend removing the v_ prefix since it's already the property of a vector object.

wcourtney reviewed (changes requested):
Looking much closer than last pass! I think the test infra could use some thought, which might affect the design. We don't have dependency injection, but hopefully you can somehow mock imports to test the behavior.

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+`Cirq <https://quantumai.google/cirq>`__ is a Python library for writing, manipulating, and optimizing quantum
+circuits and running them against quantum computers and simulators.
+
+This module is **cirq-web**, which allows users to take advantage of Typescript-based 3d visualization tools
```
👍

seunomonije replied:
Done

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere_test.ts:
```diff
@@ -0,0 +1,37 @@
+import {expect} from 'chai';
+import {BlochSphere} from './bloch_sphere';
+
+describe('The BlochSphere class', () => {
```
See new testing frmwrk in next batch of commits.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere.ts:
```diff
@@ -0,0 +1,96 @@
+
+  /**
+   * Returns the radius of the bloch_sphere.
+   * Used for testing purposes.
```
See new testing framework

seunomonije commented on cirq-web/cirq_ts/e2e/bloch_sphere/test_bloch_sphere.ts:
```diff
@@ -0,0 +1,89 @@
+      await browser.close();
+    });
+
+    it('Bloch sphere with |0⟩ statevector is correct', () => {
```
Done

seunomonije commented on cirq-web/cirq_ts/e2e/bloch_sphere/test_bloch_sphere.ts:
```diff
@@ -0,0 +1,89 @@
+      await browser.close();
+    });
+
+    it('Bloch sphere with |0⟩ statevector is correct', () => {
```
Done

seunomonije commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+    sphere = BlochSphere()
+    display(sphere)
+
+Note that you can pass a state vector into the :code:`BlochSphere()` constructor to view a particular
```
My thought process is that since this is the PyPI readme, and in production users won't be accessing the typescript at all, its safe to provide examples using just the Python side.

balopat commented on cirq-web/cirq_ts/dist/index.html:
```diff
@@ -0,0 +1,13 @@
+  <div id="container"></div>
+  <script src="bloch_sphere.bundle.js"></script>
+  <script>
+    CirqTS.blochSphere('{"radius": 5}');
```
This shouldn't be CirqTS though, I think this would read better as `BlochSphere.create(...)`

seunomonije replied:
CirqTS is the [bundle's library name](https://github.com/seunomonije/Cirq/blob/d59d56aa67cdd2bf7ff922d2c91dc83938bd2937/cirq-web/cirq_ts/webpack.config.js#L30), as we add more files I was thinking that we could have something like `CirqTS.3DCircuit(...)`, `CirqTS.LatticeViz(...)`, etc. But as we add more files to the bundle every function in each visualizations `main.ts` would be accessible by calling `CirqTS.functionName()`.

seunomonije replied:
I also noticed the comments at the top of the linked `webpack.config.js` and fixed them locally.

balopat replied:
Oh, but then the bundle is called  `<script src="bloch_sphere.bundle.js"></script>`. I thought we'll have separate bundles...or all bundles will create the same library name? That sounds like they might conflict with each other?

seunomonije replied:
In my head all the bundles would be under the same library name, and we would deal with conflicts like shown [here](https://webpack.js.org/configuration/output/#outputlibrary). If we import only one bundle it shouldn't be a problem, if we import multiple than we can specify. I can also clean up the config to reflect this

seunomonije replied:
I actually thought about this more and I actually agree with having no name for the library and just relying on referring to the `bundle.js`...changing this now

balopat commented on cirq-web/cirq_ts/package.json:
```diff
@@ -0,0 +1,65 @@
+  "scripts": {
+    "start": "webpack serve --mode development",
+    "build": "npm test && webpack --mode production",
+    "test": "webpack --mode production && mocha -r ts-node/register 'e2e/**/*.ts' 'src/**/*_test.ts'",
```
1. we want to separate unit tests from e2e tests, because unit tests are faster.
2. also, `webpack --mode production` can be replaced by `npm build`

balopat commented on cirq-web/cirq_ts/package.json:
```diff
@@ -0,0 +1,65 @@
+    "start": "webpack serve --mode development",
+    "build": "npm test && webpack --mode production",
+    "test": "webpack --mode production && mocha -r ts-node/register 'e2e/**/*.ts' 'src/**/*_test.ts'",
+    "unittest": "mocha -r ts-node/register 'src/**/*_test.ts'",
```
ah, I see you have a separate unittest...maybe we could do `test` for unit-tests and `test-all` for the above...

seunomonije replied:
![image](https://user-images.githubusercontent.com/17647506/123001093-f893c400-d375-11eb-8d20-156f7e8486d0.png) Locally I have this.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/main.ts:
```diff
@@ -0,0 +1,45 @@
+ * @param containerId A string containing the container (div, span, etc.) id that will contain the visualization
+ * output.
+ */
+export function blochSphere(
```
I find the signature of this function confusing: 
- `circleData` - there is no circle - a better could be `blochSphereConfigJSON`? 
- `vectorData` - what vector? Typically you want to name things closer to their domain name - this case it is a _state vector_. 
- `data` is superflous, maybe you can use `JSON` to signify the format

The name of the method is `blochSphere` but it does two things: it creates a bloch sphere and then adds a vector to it. Consider one of the following options:
- rename to `renderBlochSphereWithStateVector()`
- rename to`renderBlochSphere()` and return the blochsphere instance on which then the client can call `blochSphere.addVector(...)`

seunomonije replied:
I like the second, specifically because it will make it easier for modify the code in the future to support adding multiple vectors cleanly (it allows for it right now, but the vectors are all the same color).

seunomonije replied:
Except wouldn't it need to be `addVector(blochSphere, stateVectorJSON)`? If we wanted create another BlochSphere class in main, we could go ahead and do that too.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/main.ts:
```diff
@@ -0,0 +1,45 @@
+  const scene = new BlochSphereScene();
+  scene.addSceneToHTMLContainer(containerId || 'container');
+
+  const bloch_sphere = new BlochSphere(sphereData.radius);
```
TypeScript should not use snake case, this should be camel case

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/main.ts:
```diff
@@ -0,0 +1,45 @@
+
+  const bloch_sphere = new BlochSphere(sphereData.radius);
+  scene.add(bloch_sphere);
+  //bloch_sphere.addToScene(scene);
```
clean up comment

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/meridians_test.ts:
```diff
@@ -0,0 +1,189 @@
+  });
+
+  describe('configurables', () => {
+    it('createHorizontalChordMeridians() generates the correct number of lines given valid numbers', () => {
```
I found a bug :) I tested with 1 (playing around with the visualization provided by Suresh) and it creates 3 horizontal lines - this test fails for 1 `lineValue`, I'd expect 1 `lineNumber`, but it gives 3. 

Also, for 4 why do we expect 5?

Same thing visually: 

![image](https://user-images.githubusercontent.com/635073/123001861-36e1b100-d37f-11eb-875b-a0951cf870b2.png)

seunomonije replied:
I outline this in the docstring for building Meridians (https://github.com/seunomonije/Cirq/blob/d59d56aa67cdd2bf7ff922d2c91dc83938bd2937/cirq-web/cirq_ts/src/bloch_sphere/components/meridians.ts#L89). 

Honestly it was a personal asethetic choice, I like a minimum of three for the horizontal chord meridians. I can add cases for 1 and 2 though if you feel strongly about it.

And in the same comments, an odd number of horizontal meridians is needed because I think asethetically there needs to be an equator meridian. Then all the other meridians can go based off of that..if the number is even, then the meridians are unbalanced.

balopat replied:
Ah, got it, that makes more sense. What you just described should be documented in the test cases and a comment would help. The tests serve as a living documentation (besides the doc string which is good for publishing stuff off of source code).

I think I do want to experiment with a single meridian given Suresh's design: 

<img src="https://user-images.githubusercontent.com/635073/123004455-9d1c0300-d382-11eb-8a09-82b01a477734.png" height='200px'>

seunomonije replied:
Bet..i'll add support for single horizontal merdians. That design looks cool too.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere_test.ts:
```diff
@@ -0,0 +1,245 @@
+  });
+});
+
+describe('Blochsphere configurables', () => {
```
```suggestion
describe('BlochSphere configurables', () => {
```

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/axes_test.ts:
```diff
@@ -0,0 +1,56 @@
+      for (const child of children) {
+        const material = child.material as LineDashedMaterial;
+        //.eql is deep equal, which can be used to compare objects
+        expect(material.color).to.eql(new Color(1, 1, 1));
```
it's not obvious to me why `#fff` equals 1...Why is that?

seunomonije replied:
I'll add a comment for it, it's just saying that the color is white. RGB for white is (1, 1, 1)

balopat replied:
Is it though? I thought that would be (255, 255, 255)

seunomonije replied:
Here's how it works in three.js (https://threejs.org/docs/#api/en/math/Color.r). If we passed it in as a string `rgb(1, 1, 1)`, then I'd think that would be true, but in the `Color` constructor the parameters are red, green, and blue values in between 0 and 1.

balopat replied:
Ahhhhhhhh, got it. My bad.

balopat replied:
Also, github renders the color using `rgb(233,210,50)`? that's cool.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/enums.ts:
```diff
@@ -0,0 +1,10 @@
+ * of meridian they are
+ */
+export enum Orientation {
+  HORIZONTAL_CHORD,
```
I think we won't ever use the `HORIZONTAL` one...we could just rename `HORIZONTAL_CHORD` to `HORIZONTAL` and then remove the other.

seunomonije replied:
Rip horizontal meridians..I thought they could be a nice bonus feature sometime. I'll take them out and all the code for it

balopat replied:
Principle: [YAGNI ](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/meridians_test.ts:
```diff
@@ -0,0 +1,189 @@
+      });
+    });
+
+    it('giving bad orientation enum in the constructor will return no circles', () => {
```
We should throw an error here instead of silently creating an empty object. The principle is "Fail fast" - let the user know that something wrong as early as possible, with a meaningful error message. Otherwise they'll have to run the whole thing, display it in a browser and then start debugging.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/assets/fonts/helvetiker_regular.typeface.json:
```diff
@@ -0,0 +1 @@
+{"glyphs":{"ο":{"x_min":0,"x_max":712,"ha":815,"o":"m 356 -25 q 96 88 192 -25 q 0 368 0 201 q 92 642 0 533 q 356 761 192 761 q 617 644 517 761 q 712 368 712 533 q 619 91 712 201 q 356 -25 520 -25 m 356 85 q 527 175 465 85 q 583 369 583 255 q 528 562 583 484 q 356 651 466 651 q 189 560 250 651 q 135 369 135 481 q 187 177 135 257 q 356 85 250 85 "},"S":{"x_min":0,"x_max":788,"ha":890,"o":"m 788 291 ...
```
I think we don't need this anymore, right?

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere_test.ts:
```diff
@@ -0,0 +1,245 @@
+      expect(sphere.radius).to.equal(3);
+    });
+
+    it('defaults correctly to an invalid radius', () => {
```
"fail fast", throw an error instead of silently creating a seemingly valid object.

balopat commented:
The latest version currently doesn't do the live reload when I change TS files...

seunomonije commented:
When I revert back to this state it's working for me, which file are you editing ?

balopat commented:
> When I revert back to this state it's working for me, which file are you editing ?

bloch_sphere.ts - I changed the meridians and I had to manually Ctrl+R to see the change. I'll do some debugging tomorrow!

balopat commented:
> > When I revert back to this state it's working for me, which file are you editing ?
> 
> bloch_sphere.ts - I changed the meridians and I had to manually Ctrl+R to see the change. I'll do some debugging tomorrow!

It's actually working, so not sure what I saw yesterday, ignore me! :)

95-martin-orion commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,91 @@
+
+*This directory contains the code and instructions for visualization tools in a web browser or Colab/Juptyer notebooks. We do this using Typescript. While necessary, this is only half of the code needed to run the `cirq-web` project. For information on how to integrate projects here with Python and the wider Cirq package, see the `cirq_web` directory.*
+
+Visualizations run on [NodeJS](https://nodejs.org/en/), and we use [npm](https://www.npmjs.com/) for package management. To start developing, clone the repository and run `npm install` within this directory to install the necessary packages and begin development. You will need to install Node and npm if you haven't already.
```
Is there any way to run the `npm` instructions from the top-level Cirq directory instead? Not mandatory, but it would be convenient.

Also may be useful to document known error messages in `npm install`. I got this:
```
4 packages are looking for funding
  run `npm fund` for details
3 moderate severity vulnerabilities
To address all issues (including breaking changes), run:
  npm audit fix --force
Run `npm audit` for details.
```
It's unclear if those messages should inspire worry.

seunomonije replied:
I think me and Balint discussed moving to the top at one point..we're opting keep NodeJS and npm stuff within `cirq_ts` since we want cirq-web in general to be an opt in feature of Cirq. I'm pretty sure if we had it in the top-level directory than you would need to have it as a dependency..correct me if there's a way around it. 

There actually is a way that you can run `npm` instructions from the top level though, if you ran `npm --prefix 'cirq-web/cirq_ts'` you should be able to interact with everything as if you were in the `cirq_ts` directory. I can add this into the documentation

95-martin-orion replied:
Including the `npm --prefix 'cirq-web/cirq_ts'` note would be nice as a convenience item.

seunomonije replied:
Done

balopat replied:
This still reads as `npm install`. I would suggest implementing `check/npm` with the prefix included.

95-martin-orion commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,91 @@
+
+### Visualization build structure
+
+The reference example for the build structure of a visualization is the Bloch sphere. Reference the `src/bloch_sphere/main.ts` file and the `CirqBlochSphere.class.ts` file to see the code. The `src/bloch_sphere/` directory should serve as a guide for how Typescript visualizations in Cirq should be structured. Visualizations should have:
```
I don't see `CirqBlochSphere.class.ts` anywhere in this directory. Maybe `bloch_sphere.ts`?

95-martin-orion commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere.ts:
```diff
@@ -0,0 +1,123 @@
+
+    this.addSphere();
+    this.addHorizontalMeridians();
+    this.addVerticalMeridians();
```
Nit: this is missing the meridian in the XZ plane.

seunomonije replied:
For the sake of the visualization aesthetic I'd prefer to keep the meridians in the two planes..thoughts?

95-martin-orion replied:
This isn't a huge sticking point for me - I'm used to [QUIRK](https://algassert.com/quirk)'s Bloch sphere visualizations, which could be argued to have either two or three meridians. OK to leave as-is.

95-martin-orion commented on cirq-web/cirq_ts/dist/index.html:
```diff
@@ -0,0 +1,14 @@
+  <script src="bloch_sphere.bundle.js"></script>
+  <script>
+    const blochSphere = renderBlochSphere('{"radius": 5, "hMeridians": 1, "vMeridians": 1}');
+    blochSphere.addVector();
```
Recommend passing in a `VectorData` value here, so the reader can see what the format looks like. This could also be presented as an opportunity for users to try out hot-reloading, e.g. by flipping the vector from |0) to |1).

seunomonije replied:
Done

95-martin-orion commented on cirq-web/cirq_ts/src/bloch_sphere/components/vector.ts:
```diff
@@ -0,0 +1,94 @@
+    } else {
+      this.x = 0;
+      this.y = 0;
+      this.z = 0;
```
Recommend _not_ using (0, 0, 0) as the default vector. The direction of the zero-vector is undefined, which means this depends on implicit behavior in three.js and may cause confusion for devs.

seunomonije replied:
Good point..fixed

95-martin-orion commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,91 @@
+
+  The screenshot of the HTML browser output must live in a temporary directory; we use the [temp](https://github.com/bruce/node-temp) package to handle that for us. Reference the test at `e2e/bloch_sphere/test_bloch_sphere.ts` to see how to easily generate the screenshot in a temporary directory.
+
+  The pre-generated PNG file is a screenshot of the developer's choice that represents what the visualization should look like. Each visualization is required to have at least one expected PNG screenshot. For more complex visualizations, multiple screenshots may be needed. It also may be helpful to include a script that automatically generates this expected image for you, see `utils/ss_bloch_sphe ...
```
Close the final inline code block: \`npx ts-node utils/ss_bloch_sphere.ts(\`)

95-martin-orion commented on cirq-web/cirq_web/README.md:
````diff
@@ -0,0 +1,80 @@
+            <script>YOUR_BUNDLE_FUNCTION<script>
+        """
+``` 
+This will allow your visualiztion to be displayed in a notebook cell with:
````
`visualiztion` -> `visualization`

95-martin-orion commented on cirq-web/example.ipynb:
```diff
@@ -0,0 +1,177 @@
+{
```
I recommend using the standard Cirq notebook header, with links to the devsite / Colab / github. An example can be seen at the top of [this file](https://github.com/quantumlib/Cirq/blob/master/docs/circuits.ipynb).

seunomonije replied:
Done

95-martin-orion commented on cirq-web/example.ipynb:
```diff
@@ -0,0 +1,177 @@
+       "\n",
+       "        <div id=\"container\"></div>\n",
+       "        <script>/*! For license information please see bloch_sphere.bundle.js.LICENSE.txt */\n",
+       "(()=>{\"use strict\";var t={212:(t,e,n)=>{n.r(e),n.d(e,{ACESFilmicToneMapping:()=>nt,AddEquation:()=>T,AddOperation:()=>Q,AdditiveAnimationBlendMode:()=>qe,AdditiveBlending:()=>b,AlphaFormat:()=>Ot,AlwaysDepth:()=>V,AlwaysStencilFunc:()=>Sn,AmbientLight:()=>Pu,AmbientLightProbe:()=>$u,AnimationClip:()=>Dh,AnimationLoader:()=>Gh,AnimationMixer:()=>Cd,AnimationObjectGroup:()=>Ld,AnimationUt ...
```
[No change needed] I wasn't able to test this in colab due to `cirq_web` being unreleased. If you'd like, I can try getting this running locally to inspect more closely.

seunomonije replied:
You can test with `jupyter notebook` if you wanna try the Jupyter notebook, also have this Colab with the PyPI package if you want to try that as well: https://colab.research.google.com/drive/1Jukad4xTwU_f9YiJ8GsPiK6BidQQSm3s?resourcekey=0-9zSsrwsyeJIxq0SV4ccqWA&usp=sharing

95-martin-orion replied:
Thanks for the link! All seems good to me there.

seunomonije commented:
I found a bug when going through everything when trying to display Bloch spheres in two different cells..since we were always adding the three.js scene to a div like so : `<div id="container"><div>`, the scenes would just stack in the same div. I fixed this by having bloch sphere be added to a class called `bloch-sphere-container`, and added logic when adding a new scene to add a scene to the first empty instance of a `bloch-sphere-container` div.

I also found a bug when building the HTML for the visualization that assigning the bloch sphere to a variable is problematic, since running code in multiple cells instantiates the variable multiple times. I fixed this by removing variable declarations entirely, and documenting it here in case anyone else in the future comes up with the problem.

The next commit has the code for both of these fixes.

wcourtney reviewed (changes requested):
Partial review, but should inform some changes in other tests as well. Overall, I'm really happy with the progress on the test infra, great progress!

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/sphere.ts:
```diff
@@ -0,0 +1,69 @@
+
+  /**
+   * Class constructor
+   * @param radius the desired radius of the sphere
```
Will follow up on this, need to a bit of research

seunomonije replied:
So from what I'm seeing units aren't straightforward in three.js. Searching for units on the three.js official website doesn't give anything (can keep working on it).

I looked at [this thread ](https://discourse.threejs.org/t/what-unit-of-distance-in-threejs/20782/12)and its saying that the units are abstract, that three.js works in world units, and also that three.js works using SI, [where one unit in three.js equals one meter](https://discoverthreejs.com/tips-and-tricks/#work-in-si-units).

A lot of conflicting information, is there a standard for things like this?

wcourtney replied:
IIUC, the size of the sphere could be radius 1, but will be various sizes on the screen depending on the distance of the camera. In the DOM, it's common to align on 'pixels' as the absolute unit of distance on the screen, and I think that makes sense for the radius here. WDYT?

[Here](https://discourse.threejs.org/t/functions-to-calculate-the-visible-width-height-at-a-given-z-depth-from-a-perspective-camera/269/19) is a case of someone trying to align pixels with the three.js world space.

@balopat  - do you have any preference whether this gets punted to a new issue vs completed now?

balopat replied:
While theoretically correct to set the radius to be 1 (as it is mathematically!) - that only works in isolation. If the Bloch sphere will be combined with gates though in the 3D circuits, this geometrical correctness will get in the way. 

I love the discourse experiment you linked - but still lives in a forum thread - taking that on will require designing it properly. Let's open an issue around this.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere.ts:
```diff
@@ -0,0 +1,123 @@
+   * Adds the Bloch sphere to the designated Three.js Scene.
+   * @param scene A Three.js scene object
+   */
+  addToScene(scene: Scene | BlochSphereScene) {
```
This can actually go..an arbitary scene object and a BlochSphere scene can support adding this with just `.add(BlochSphere)`

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/components/axes.ts:
```diff
@@ -0,0 +1,126 @@
+  /**
+   * Class constructor.
+   * @param radius The radius of the circle the axes will belong to. This should be the same as the sphere.
+   * @param xAxisColor The color of the x axis as a string.
```
I think the overall sentiment of the TS should be make things as easy to customize as possible so that working with it from the Python/HTML side is as easy as possible. I think the same question is valid for the number of meridians displayable on the Bloch sphere, the radius of the circle, etc. That being said, I agree and don't see a need for per-axis customization.

I think this is something that we should probably introduce in the next PR with the design fixes. I like the enum for color schemes too, we could have a dark mode and similar themes.

wcourtney replied:
Yeah, I really meant "are you specifying these parameters anywhere, or just using the defaults"? I strongly recommend a YAGNI approach when designing interfaces. It will make it simpler to understand how the feature is intended to be used. It also means that you don't need to test all of the ways that a user might pass you bad data ;-). It sounds like these should just be constants for the application/class rather than parameters.

seunomonije replied:
Ah ok I see now..changing

95-martin-orion reviewed (commented):
No further issues from me! Leaving final approval to Will / Balint.

balopat commented on .github/workflows/ci.yml:
```diff
@@ -290,3 +290,36 @@ jobs:
+      - name: Lint
+        run: check/ts-lint 
+  ts-test:
+    name: Typescript unit testing
```
This is not just unit test, but also the e2e tests, let's call this `Typescript tests`.

balopat commented on .gitignore:
```diff
@@ -48,4 +48,4 @@ docs/api_docs
 .python-version
 
 # notebook test output
-out
```
ping, can we remove this change?

seunomonije replied:
I don't know what's going on with this, I've added new/removed new lines, even copied the entire .gitignore file in and it's still saying that there's something wrong with it

balopat replied:
Try `git checkout master .gitignore` and then `git commit -am "revert .gitignore"`. 
It is probably your editor that is adding/removing last lines.

balopat commented on check/ts-build:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
nit: add copyright notice

balopat commented on check/ts-build:
```diff
@@ -0,0 +1,15 @@
+cd "$(git rev-parse --show-toplevel)"
+
+cd 'cirq-web/cirq_ts'
+npx webpack --mode production
```
we typically forward the arguments to scripts. Add `$@` here. 
You can test whether it works by `check/ts-build --watch` for example.

balopat commented on check/ts-coverage:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
nit: add copyright notice

balopat commented on check/ts-coverage:
```diff
@@ -0,0 +1,14 @@
+cd "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+cd "$(git rev-parse --show-toplevel)"
+
+npm --prefix 'cirq-web/cirq_ts' run coverage
```
nit: forward arguments!

balopat commented on check/ts-lint:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
nit: add copyright notice

balopat commented on check/ts-lint:
```diff
@@ -0,0 +1,14 @@
+cd "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+cd "$(git rev-parse --show-toplevel)"
+
+npm --prefix 'cirq-web/cirq_ts' run lint
```
forward arguments

balopat commented on check/ts-lint-and-format:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
nit: add copyright notice

balopat replied:
Shebang line [should be always on the first line](https://stackoverflow.com/questions/12910744/why-should-the-shebang-line-always-be-the-first-line#:~:text=4%20Answers&text=The%20shebang%20must%20be%20the,file%20available%20to%20that%20program.).

balopat commented on check/ts-lint-and-format:
```diff
@@ -0,0 +1,14 @@
+cd "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+cd "$(git rev-parse --show-toplevel)"
+
+npm --prefix 'cirq-web/cirq_ts' run fix
```
forward args

balopat commented on check/ts-test:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
nit: add copyright notice

balopat replied:
Shebang line [should be always on the first line](https://stackoverflow.com/questions/12910744/why-should-the-shebang-line-always-be-the-first-line#:~:text=4%20Answers&text=The%20shebang%20must%20be%20the,file%20available%20to%20that%20program.).

balopat commented on check/ts-test:
```diff
@@ -0,0 +1,16 @@
+cd "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+cd "$(git rev-parse --show-toplevel)"
+
+npm --prefix 'cirq-web/cirq_ts' run test
```
forward args

balopat commented on .github/workflows/ci.yml:
```diff
@@ -290,3 +290,36 @@ jobs:
+        with:
+          node-version: '14.16.1'
+      - name: Install npm dependencies
+        run: npm --prefix 'cirq-web/cirq_ts' ci
```
refactor `npm --prefix 'cirq-web/cirq_ts'` to a new script `check/npm` which forwards all the arguments and replace the original call in all the scripts, so that in case the prefix changes, we can change it in a single place.

balopat commented on cirq-web/MANIFEST.in:
```diff
@@ -0,0 +1 @@
+include cirq_ts/dist/*.bundle.js
```
we don't need this file, MANIFEST.in is only for source distributions and we don't publish source distros.

seunomonije replied:
Sounds good, I fixed by just adding to package_data

balopat replied:
Yes, package_data determines what the wheel file (binary distribution) contains.
I guess you still need to delete this file! :)

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+
```
nit: remove new lines on the top

balopat replied:
We should also add the Cirq logo here. This is the description that gets published to PyPi.

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+
+Documentation
+-------------
+All current documentation for cirq-web can be found in the README files located in this module's subdirectories.
```
```suggestion
Documentation for cirq-web can be found in the README files located in this module's subdirectories.
```

Also, link to the repo.

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+-------------
+All current documentation for cirq-web can be found in the README files located in this module's subdirectories.
+
+Below is a quick example of how to generate a portable 3d rendering of a bloch sphere using cirq-web:
```
```suggestion
Below is a quick example of how to generate a portable 3D rendering of the Bloch sphere using cirq-web:
```

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+`Cirq <https://quantumai.google/cirq>`__ is a Python library for writing, manipulating, and optimizing quantum
+circuits and running them against quantum computers and simulators.
+
+This module is **cirq-web**, which allows users to take advantage of browser based 3d visualization tools
```
```suggestion
This module is **cirq-web**, which allows users to take advantage of browser based 3D visualization tools
```

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+    sphere.generate_html_file()
+
+This will create the file in the current working directory. There are additional options to specify the
+output directory, open the visualization in a browser, etc. 
```
```suggestion
output directory or to open the visualization in a browser for example. 
```

balopat commented on cirq-web/README.rst:
```diff
@@ -0,0 +1,48 @@
+This will create the file in the current working directory. There are additional options to specify the
+output directory, open the visualization in a browser, etc. 
+
+You can also view and interact with a bloch sphere in a Colab or Jupyter notebook setting
```
```suggestion
You can also view and interact with a Bloch sphere in a Colab or Jupyter notebook setting
```

balopat commented on cirq-web/cirq_ts/.eslintrc.json:
```diff
@@ -0,0 +1,12 @@
+      "files": ["*.ts"],
+      "rules": {
+        "node/no-unpublished-import": "off",
+        "max-len": [2, 120]
```
can we make this 100? It is 100 in python.

seunomonije replied:
Sounds good.

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+## Cirq Typescript Development
+
+*This directory contains the code and instructions for visualization tools in a web browser or Colab/Juptyer notebooks. We do this using Typescript. While necessary, this is only half of the code needed to run the `cirq-web` project. For information on how to integrate projects here with Python and the wider Cirq package, see the `cirq_web` directory.*
```
nit: link to `cirq_web` directory

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+Visualizations run on [NodeJS](https://nodejs.org/en/), and we use [npm](https://www.npmjs.com/) for package management. To start developing, clone the repository and run `npm install` within this directory to install the necessary packages and begin development. You will need to install Node and npm if you haven't already.
+
+
+For developing 3D visualizations, we rely on the [three.js](https://threejs.org/) framework. Understanding their documentation is critical to understanding 3D visualizations in Cirq. 
```
```suggestion
For developing 3D visualizations, we rely on the [three.js](https://threejs.org/) framework. 
```
This sentence is unnecessary - people can prioritize their level of involvement themselves in the threejs docs, our job is to point them to the places that they can use.

balopat commented on cirq-web/cirq_ts/README.md:
````diff
@@ -0,0 +1,96 @@
+
+As an additional note, all `npm` commands can be ran from the top-level of Cirq like so:
+```
+npm --prefix 'cirq-web/cirq_ts' [YOUR_COMMAND]
````
replace with explaining what `check/npm` does.

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,77 @@
+A good example on the build structure of a visualization is the Bloch sphere. Reference the `src/bloch_sphere/main.ts` file and the `CirqBlochSphere.class.ts` file to see the code. The `src/bloch_sphere/` directory should serve as a guide for how Typescript visualizations in Cirq should be  
+developed in general. In the general sense, visualizations should have:
+ - A "root" folder within the `src/` directory labeled according to the title of the visualization. All files and directories for a particular visualization will live here. In the case of the Bloch Sphere, this is `bloch_sphere/`.
+ - A `components/` directory which contains classes representing different components of the larger visualization, following typical object oriented programming techniques. In the case of the Bloch sphere, you can see that we have different classes for `Axes`, `Meridians`, `Text` etc.
```
+1 let's not forget to add an issue for this.

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+
+### Creating visualization bundle files
+
+Following this structure, you will be able to bundle your visualization by running the command `npx webpack --mode production`. This will build the bundled Javascript file(s) within the `dist/` directory, where you can access and reference them in HTML.
```
I guess we'll need a `check/npx` as well

seunomonije replied:
Right now I go into the `cirq_ts` directory and run, I don't think `--prefix` is supported yet for it, once I'm done i'll do some more digging

https://github.com/npm/npx/issues/74
https://stackoverflow.com/questions/60739550/how-to-specify-pathlike-npm-prefix-in-npx

balopat replied:
OK, in that case you can simulate that with 
```
#!/bin/bash

cd cirq-web/cirq_ts
npx $@
```

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+
+There are two main ways to develop visualizations you are creating in Cirq. The first, and recommended way, is to spin up a Webpack development server and view your visualizations in the browser. You can also develop using Jupyter notebook if you want to easily test integration with Python code. 
+
+#### Hot reloading development server (Recommended)
```
nit: add newline below

balopat commented on cirq-web/cirq_ts/README.md:
````diff
@@ -0,0 +1,96 @@
+  ...
+};
+```
+Note that the bundled files that `webpack-dev-server` creates live in memory, so you won't be able to see them. However, you can access them like so:
````
```suggestion
Note that the bundled files that `webpack-dev-server` creates live in memory, so you won't be able to find them on the file system.
```

balopat commented on cirq-web/cirq_ts/README.md:
````diff
@@ -0,0 +1,96 @@
+```
+Note that the bundled files that `webpack-dev-server` creates live in memory, so you won't be able to see them. However, you can access them like so:
+```html
+<script src="YOUR_VIZ_NAME.bundle.js"></script>
````
```suggestion
<script src="/YOUR_VIZ_NAME.bundle.js"></script>
```

Right? It is served on the root of the context. 
We can also mention in addition that from the browser `http://localhost:8080/YOUR_VIZ_NAME.bundle.js` is a way to inspect the bundled js.

balopat commented on cirq-web/cirq_ts/README.md:
````diff
@@ -0,0 +1,96 @@
+<script src="YOUR_VIZ_NAME.bundle.js"></script>
+```
+
+### Developing in a Jupyter Notebook
````
nit: add new line below

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+
+**NOTE:** In order to access the bundled javascript, you need to include the full path to it: `cirq_ts/dist/YOUR_VIZ_NAME.bundle.js`. If you make any changes to the directory structure, take into account that the path may change as well.
+
+### Developing in Google Colaboratory
```
nit: newline

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+### Developing in Google Colaboratory
+We currently do not support developing visualizations in Google Colaboratory notebooks. However, visualization ran from the PyPI package are able to be viewed in Colab. 
+
+### Formatting and linting
```
nit: newline

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+### Formatting and linting
+All Typescript files need to be formatted/linted according to [Google's public Typescript style guide](https://google.github.io/styleguide/tsguide.html). We use (Google's open source tool GTS)[https://github.com/google/gts] to handle this for you. Run `npm run fix` to handle fixing changes automatically, or refer to the `package.json` file for more options.
+
+### Testing
```
nit: newline

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+
+We use [Mocha](https://mochajs.org) and [Chai](https://www.chaijs.com/) as our main testing tools. For comparing image diffs, we use [Pixelmatch](https://github.com/mapbox/pixelmatch) and [pngjs](https://github.com/lukeapage/pngjs).
+
+#### Unit testing
```
nit: newline

balopat commented on cirq-web/cirq_ts/README.md:
```diff
@@ -0,0 +1,96 @@
+
+Unit tests must live adjacent to their source file with the `_test.ts` suffix. So for the file `dir/MyFile.ts`, the corresponding testfile will be `dir/MyFile_test.ts`.
+
+#### Visualization testing
```
nit: newline

balopat reviewed (changes requested):
Bunch of nits.
+ 1 more nit: Add copyright header to cirq-web/cirq_ts/__init__.py

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/bloch_sphere_e2e_test.ts:
```diff
@@ -0,0 +1,140 @@
+    </script>
+    </body>
+</html>
+`;
```
nit: dedupe these two strings with Javascript's string interpolation.

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/bloch_sphere_e2e_test.ts:
```diff
@@ -0,0 +1,140 @@
+      await browser.close();
+    });
+
+    it('Bloch sphere with custom statevector matches the standard PNG copy', () => {
```
```suggestion
    it('Bloch sphere with custom statevector matches the gold PNG', () => {
```

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/bloch_sphere_e2e_test.ts:
```diff
@@ -0,0 +1,140 @@
+        {threshold: 0.1}
+      );
+      // Accouting for the difference in unicode symbols in browsers
+      expect(pixels).to.be.below(700);
```
I'm not sure I get this - I think this should be 0, can you explain a bit more what's happening?

seunomonije replied:
This was something I explored and then forgot about...I just fixed it. Good lesson don't change the tests.

So Chromium is weird..what I think happened is that when the browser instance starts, a alert comes from the top of the screen that says "Chromium is being ran under testing" or something and then when puppeteer takes a screenshot, it's getting that image along with the bloch sphere. I was failing my tests a while ago due to it and looked at example screenshots and thought the symbols looked different in different browsers so documented it and then moved on. I adding an argument to the `puppeteer.launch()` function, and now it works with equal 0.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere_test.ts:
```diff
@@ -0,0 +1,281 @@
+    child => child.constructor.name === 'Labels'
+  ) as Labels;
+
+  it('adds a single object to a scene of type Group, specifically BlochSphere', () => {
```
```suggestion
  it('adds a single object to a scene of type Group, specifically BlochSphere', () => {
```
```suggestion
  it('adds a single BlochSphere to a scene of type Group', () => {
```

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/axes.ts:
```diff
@@ -0,0 +1,103 @@
+   * @param halfLength The halfLength of the axes line. This should be the same as the radius of the sphere.
+   * @param xAxisColor The color of the x axis as a string.
+   * @param yAxisColor The color of the y axis as a string.
+   * @param zAxisColor The color of the z axis as a string.
```
these colors are now not used anywhere, remove them from the docstring.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/scene.ts:
```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 The Cirq Developers
```
How are we testing this file? How is it covered?

balopat commented on cirq-web/cirq_ts/dist/index.html:
```diff
@@ -0,0 +1,14 @@
+<!doctype html>
```
I think this should be a front page for all the development files, i.e. a "table of contents" or "index" for "bloch sphere", "3d circuit type 1", "3d circuit type 2" , etc.  - but we can do that in the next PR as we add more widgets.

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/bloch_sphere_e2e_test.ts:
```diff
@@ -0,0 +1,140 @@
+// Automatically track and cleanup files on exit
+temp.track();
+
+describe('Check Bloch Sphere looks correct', () => {
```
nit: refactor the `describe` / `it` sections to describe(Bloch sphere), it(matches the default state)

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/sphere.ts:
```diff
@@ -0,0 +1,69 @@
+        'The radius of a Sphere must be greater than or equal to 1'
+      );
+    } else {
+      return Math.ceil(radius);
```
Why do we need to do ceil? Fractional units don't work? They seem to work for me.

balopat replied:
In that case we don't need "sanitize" you can also just remove this function and do the bound checking in the constructor.

seunomonije replied:
It was more of a design choice on my end, but took it out

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/vector.ts:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * Generates a vector starting at (0, 0) and ending at the coordinates
+   * of the given parameters, and adds to group.
+   * Utilizes three.js ArrowHelper function generate the vector.
```
These arguments are odd: x, y, z should unambiguously determine length of a vector just from the Pythagorean theorem.  

I understand that this came from the `cirq.bloch_vector_from_state_vector` as that gives the coordinates of a pure state on a sphere of radius 1, but that doesn't mean that it should influence the API in surprising ways.

seunomonije replied:
Three.js helps out with a distance calculation function..I removed code to normalize the vector and get the length of the vector based on the coordinates

balopat commented on cirq-web/cirq_ts/utils/ss_bloch_sphere.ts:
```diff
@@ -0,0 +1,52 @@
+// Copyright 2021 The Cirq Developers
```
do we need this file and the utils folder? 
It seems like this was an experiment for the e2e tests.

seunomonije replied:
I had it as a utility for adding new screenshots for e2e tests as you updated the bloch sphere, but I also think that's something that users can do themselves.

balopat commented on cirq-web/cirq_web/_version.py:
```diff
@@ -0,0 +1,18 @@
+# limitations under the License.
+
+"""Define version number here, read it from setup.py automatically,
+and warn users that the latest version of cirq uses python 3.6+"""
```
There is no warning about python 3.6 here - remove copy pasta!

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,158 @@
+        self,
+        sphere_radius: int = 5,
+        state_vector: ndarray = to_valid_state_vector([math.sqrt(2) / 2, math.sqrt(2) / 2]),
+        random: bool = False,
```
Let's remove randomness from everywhere, while it was good to experiment with it for the prototypes, it is not really useful for an end user.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,158 @@
+    def __init__(
+        self,
+        sphere_radius: int = 5,
+        state_vector: ndarray = to_valid_state_vector([math.sqrt(2) / 2, math.sqrt(2) / 2]),
```
We should not give a default state vector, or make it `None` instead.
Also, type should be `cirq.qis.STATE_VECTOR_LIKE`.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,158 @@
+            operating system.
+
+        For now, if ran in a notebook, this function just returns. Support for downloading
+        the HTML file via the browser can be added later.
```
No need to promise future functionality in the docstring for users. Please remove.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,158 @@
+
+    def _convert_sphere_input(self, radius: int) -> str:
+        if radius <= 0:
+            raise (BaseException('You must input a positive radius for the sphere'))
```
We never raise `BaseException`. Typically `ValueError` is raised when the provided arguments fail validation.

balopat replied:
Also, outer parenthesis are not required.

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,158 @@
+
+        super().__init__('cirq_ts/dist/bloch_sphere.bundle.js')
+
+        self.sphere_json = self._convert_sphere_input(sphere_radius)
```
This is a bigger change, but will simplify the code a bunch - I don't think we need the JSON layer here. We can just directly set the arguments in the function signature for a given object. My bad, I recommended JSON at the start of the project, but in fact JSON will be only required when we want to represent more complicated objects to pass down (maybe in the 3D Circuit case).

seunomonije replied:
Done

balopat reviewed (changes requested):
Added bunch of nits, the biggest one is that I think we should remove JSON as a middle layer - as it seems like a bit of over-engineering.

balopat commented on cirq-web/cirq_ts/dist/bloch_sphere.bundle.js:
```diff
@@ -0,0 +1,2 @@
+/*! For license information please see bloch_sphere.bundle.js.LICENSE.txt */
```
This change reminded me: as the bundles are generated, we should add a check that ensures that the generated files are up to date. I.e. 

1. `check/npm run build`
2. check if there are any changes in the dist directory - if there are, fail with a nice error message that tells the developer what to run and commit.

seunomonije replied:
So basically this is just making sure that people commit their changed bundle.js files?

seunomonije replied:
Done

balopat replied:
I don't see it added to the ci.yml - am I missing something?

balopat commented on cirq-web/cirq_ts/e2e/bloch_sphere/bloch_sphere_e2e_test.ts:
```diff
@@ -0,0 +1,130 @@
+      const page = await browser.newPage();
+
+      // Take a screenshot of the second image, adding the vector
+      browserStrings[1] += '.addVector(1, 1, 2);';
```
Nice, at this point you have a templating solution, using arrays. I would make this one tad more readable:

define a function named `function htmlContent(clientCode){}` that would inject the `clientCode` string into that middle script tag and return the whole string. Then you can use that below simply: 

```javascript
await page.setContent(htmlContent("renderBlochSphere('container').addVector(1,1,2)"))
```

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/text_test.ts:
```diff
@@ -0,0 +1,43 @@
+  it('successfully generates an arbitrary label at the correct position', () => {
+    const label = labels.children[0] as Label;
+    expect(label.position).to.eql(new Vector3(0, 0, 0));
+    // Need to figure out how to extract text from .fillText() on canvas.
```
Mark as `TODO(#<issue>)` if we plan on doing this after the PR. Let's not forget it in here.
I do think we should fix this though before the PR goes in.

seunomonije replied:
I added a property to Label to match with the structure of other tests, you can't recover the string once you draw on the canvas because it's stored as an image, another option is adding a property to the HTML element but three.js also throws away the HTML element when rendering.

balopat commented on cirq-web/cirq_ts/src/bloch_sphere/components/vector.ts:
```diff
@@ -0,0 +1,95 @@
+
+  /**
+   * Class constructor.
+   * @param x the x coordinate of the vector tip
```
It is a bit still unclear to me from this dosctring what the x,y,z are for. Describe what exactly these are...or maybe we can rename the class to `ScaledBlochVector` or `StateVector` and then it become a bit clearer, e.g. "x,y,z provide the coordinates of the state vector within the Bloch sphere with a unit radius. However, as the actual sphere's radius can be any ThreeJS unit, we can scale the vector along with it, by specifying `bloch_sphere_radius`" -- and then I'd rename the scaling_factor to `bloch_sphere_radius`.

seunomonije replied:
Changed to StateVector and added clarification

balopat commented:
Yeah, to keep the source and the generated files consistent. You generally
have two options: keep only the source in git and generate the files at
release time, or store both of them in git but then we need an extra check
to validate that they are in fact consistent.

On Thu, Jul 1, 2021 at 3:11 PM Seun Omonije ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cirq-web/cirq_ts/dist/bloch_sphere.bundle.js
> <https://github.com/quantumlib/Cirq/pull/4171#discussion_r662532359>:
>
> > @@ -0,0 +1,2 @@
> +/*! For license information please see bloch_sphere.bundle.js.LICENSE.txt */
>
> So basically this is just making sure that people commit their changed
> bundle.js files?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quantumlib/Cirq/pull/4171#discussion_r662532359>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE3BQPKYZ2IEYWRWPF2AMTTVS4XTANCNFSM46IZY3QQ>
> .
>

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,119 @@
+
+        # Generate a unique UUID for every instance of a Bloch sphere.
+        # This helps with adding visualizations to scenes, etc.
+        self.id = str(uuid.uuid1())
```
This should live in `Widget` I think. And then we can have the whole div + bundle part in there as well. The widget can provide purely the client code.

balopat reviewed (changes requested):
- [x] in all the check scripts please make sure that the Shebang line is [always on the first line](https://stackoverflow.com/questions/12910744/why-should-the-shebang-line-always-be-the-first-line#:~:text=4%20Answers&text=The%20shebang%20must%20be%20the,file%20available%20to%20that%20program.) 
- [x] MANIFEST.in is still to be removed
- [x] add check/ts-build-current to the ci.yml as a job
- [x] follow-up on Will's comment on unit for the radius
- [x] rename vector.ts to state_vector.ts
- [x] had a bunch of comments on how we handle the dist path 
- [x] had a bunch of comments on self.id 
- [ ] refactor bundle + div HTML code could live in the Widget class and the BlochSphere can just provide a client javascript code

balopat commented:
Great progress! I think these are my last comments :D "famous last words"

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,103 @@
+            The path of the HTML file in as a Path object.
+        """
+
+        client_code = f"""
```
This is now yelling abstract class pattern: 

- have an abstract method on Widget called `client_code()` - which all Widget implementors have to implement
- move the `generate_html_file` method to Widget, which will call this abstract method

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,101 @@
+    return file_path
+
+
+class Widget:
```
This should be an abstract class, there's not much value in being able to instantiate an empty widget

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,101 @@
+    return bundle_html
+
+
+def write_output_file(output_directory: str, file_name: str, contents: str) -> Path:
```
let's get rid of this method - it is very easy to write it in client code and really not typical to pull this out to a separate method: 

e.g you can write this in generate_html_file:
```
path_of_html_file = os.path.join(output_directory, file_name)
        with open(path_of_html_file) as f:
            f.write(contents)
```

balopat commented:
OK, one last push: 
- [ ] make the Widget class abstract and to use the [Template method pattern ](https://en.wikipedia.org/wiki/Template_method_pattern) (I do recommend also reviewing the gang of four [design patterns](https://en.wikipedia.org/wiki/Design_Patterns)!)
- [ ] remove the write output file method

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,101 @@
+        """Returns the bundle script of a widget"""
+        return _to_script_tag(self.bundle_file_path)
+
+    def create_html_content(self, client_code: str) -> str:
```
this can be private

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,101 @@
+        # This helps with adding visualizations to scenes, etc.
+        self.id = str(uuid.uuid1())
+
+    def get_bundle_script(self):
```
this can be private

balopat replied:
+ type hints!

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,101 @@
+        Args:
+            bundle_file_path: The relative path of the widget's bundle file starting from cirq_ts/
+        """
+        self.bundle_file_path = f'{_DIST_PATH}/{bundle_file_path}'
```
use os.path.join() instead of hardcoding the separator! Also you're mixing pathlib.Path with str...

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,101 @@
+class Widget:
+    """Parent class for all widgets."""
+
+    def __init__(self, bundle_file_path: str):
```
Now that we fixed the directory - this can be just a file name...

```suggestion
    def __init__(self, bundle_filename: str):
```

But realistically, this is not something the user has to specify ever, instead it will be governed by the implementing subclass - e.g. we will always use the same exact file path for BlochSphere - again this is a good candidate for a template method pattern - define an abstract method here which you can use in this class' private methods.

balopat commented on cirq-web/cirq_web/widget_test.py:
```diff
@@ -0,0 +1,29 @@
+# limitations under the License.
+
+import cirq_web
+
```
we need a bit more tests as we move things to Widget: create a subclass test widget and verify the behaviors - you can probably move a bunch of the test logic you're currently using for BlochSphere.

wcourtney reviewed (commented):
LGTM % some nits and the radius unit issue which you _could_ punt to an issue. Deferring to @balopat for final approval.

seunomonije commented on cirq-web/cirq_ts/src/bloch_sphere/bloch_sphere_test.ts:
```diff
@@ -0,0 +1,264 @@
+    });
+
+    it('axes lines are the right colors by default', () => {
+      expect(axes.xAxisColor).to.equal('#1f51ff');
```
Checked to make sure the colors are different

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,123 @@
+
+
+class Widget(ABC):
+    """Abstract template class for all widgets."""
```
```suggestion
    """Abstract class for all widgets."""
```

balopat commented on cirq-web/cirq_web/widget.py:
```diff
@@ -0,0 +1,123 @@
+    OTHER = 3
+
+
+def _to_script_tag(bundle_filename: str) -> str:
```
move this to the bottom of the file

balopat commented on cirq-web/cirq_web/widget_test.py:
```diff
@@ -0,0 +1,71 @@
+import cirq_web
+
+
+class MockWidget(cirq_web.Widget):
```
```suggestion
class TestWidget(cirq_web.Widget):
```

Mock evokes different concepts - like capturing call arguments, using them as an intermediary...I'd just call it `TestWidget`

seunomonije replied:
TestWidget gives a pytest warning..got stuck wondering why for like 30 mins lol, but its because if the name starts with Test pytest doesn't skip it

balopat commented on cirq-web/cirq_web/bloch_sphere/bloch_sphere.py:
```diff
@@ -0,0 +1,56 @@
+from cirq.qis.states import STATE_VECTOR_LIKE
+
+
+class BlochSphere(widget.Widget):
```
looks great now!

balopat reviewed (approved):
LGTM with some final tiny nits

balopat commented:
I think the automerge bot might be dying on this PR due to its size :)
